tests/test_core.py

```diff
@@ -714,3 +714,25 @@ def test_scalar_promote(arrays, scalar):
 def test_promotion_failures(arrays, scalar):
     with pytest.raises(TypeError, match="Cannot promote"):
         promote(*arrays, scalar)
+
+
+@pytest.mark.parametrize(
+    "x, y",
+    [
+        (ndx.asarray([1, 2, 3], dtype=ndx.int64), 1.12),
+        (1.23, ndx.asarray([1, 2, 3], dtype=ndx.int64)),
+        (True, ndx.asarray([1, 2, 3], dtype=ndx.int8)),
+        (ndx.asarray([True, False]), 1.12),
+        (ndx.asarray([True, False]), 4),
+        (ndx.asarray([1.23, 2.34]), True),
+        (ndx.asarray([1.23, 2.34]), 2),
+    ],
+)
+def test_cross_kind_promotions(x, y):
+    onnx_result = x + y
```

Comment:
Recurrence of a theme: I'd feel better if we wouldn't thread the expected NumPy result through ndonnx. The way it stands now, we are also relying on `to_numpy()` and `asarray` to do the right thing with respect to data types. Why not make `x`, and `y` numpy rather than ndonnx arrays and then conditionally convert them with `ndx.asarray` below, instead.

Reply:
I understand what you are saying but in practice you can rely on them as you would be seeing almost every test fail (including the Array API ones) if they did the wrong thing. I can change it if you insist but don't think we should get hung up on this.